File: ts-plugins/src/eventstore/compensation-service.ts
import type { VersionedTransitionEvent, CompensationResolver } from './types.js';
import type { EventLogStoreDecorator } from './event-log-store-decorator.js';

/**
 * Compensation service — records compensation events for failed transitions.
 */
export class CompensationService {
  constructor(
    private readonly resolver: CompensationResolver,
    private readonly eventLog: EventLogStoreDecorator,
  ) {}

  compensate(event: VersionedTransitionEvent, cause: Error): boolean {
    const plan = this.resolver(event, cause);
    if (!plan) return false;
    this.eventLog.appendCompensation(
      event.flowId, plan.action,
      JSON.stringify(plan.metadata),
    );
    return true;
  }
}

File: java-core-plugins/src/main/java/org/unlaxer/tramli/plugins/eventstore/CompensationService.java
package org.unlaxer.tramli.plugins.eventstore;

import java.util.Optional;

public final class CompensationService {
    private final CompensationResolver resolver;
    private final EventLogStoreDecorator eventLog;

    public CompensationService(CompensationResolver resolver, EventLogStoreDecorator eventLog) {
        this.resolver = resolver;
        this.eventLog = eventLog;
    }

    public boolean compensate(VersionedTransitionEvent sourceEvent, Throwable cause) {
        Optional<CompensationPlan> plan = resolver.resolve(sourceEvent, cause);
        if (plan.isEmpty()) {
            return false;
        }
        CompensationPlan value = plan.get();
        eventLog.recordCompensation(sourceEvent.flowId(), value.compensationName(), value.metadata());
        return true;
    }
}
